src/targets/gpu/fuse_mlir.cpp

```diff
@@ -619,68 +630,62 @@ struct find_mlir_fused_ops
                return i != x_ins and reaches(gemm_based_op, i);
            }))
             return;
-        auto names = pm->get_parameter_names();
-        std::sort(names.begin(), names.end());
+
+        std::unordered_map<instruction_ref, instruction_ref> map_ins;
         module_ref mm = mpm.create_module("mlir_" + pm->name());
         mm->set_bypass();
-        auto [anchor_op, top_inputs] = fuse_input_ops_and_gemm_based_op(
-            mm, gemm_based_op->inputs(), gemm_based_op->get_operator());
-        std::unordered_map<instruction_ref, instruction_ref> param_map =
-            create_param_map_with_literals(mm, pm, pw_ins->get_shape());
-        auto [upper_input, op_stream] = get_fusable_input_op_stream(x_ins);
-        assert(upper_input == gemm_based_op);
-        auto prev_input = anchor_op;
-        for(const auto& op : reverse(op_stream))
-        {
-            prev_input = mm->add_instruction(op, {prev_input});
-        }
-        assert(prev_input->get_shape().lens() == x_ins->get_shape().lens());
-        param_map[x_ins] = prev_input; // this is to avoid adding parameter for gemm/conv reshaped
-                                       // input to pointwise in new fused module
+        fuse_input_ops(mm, gemm_based_op->inputs(), &map_ins);
+
         bool gemm_has_multi_outs = gemm_based_op->outputs().size() > 1;
```

Comment:
Might be tangential to this PR, but what happens if one of the reshape ops has multi outputs? ie. something like:
```
dot -> reshape -> pointwise -> .... -> dot
          |_____________________________|
```

Reply:
The loop below checks that all the reshapes between dot and pointwise have only one output. For the aux inputs to pointwise, it doesnt really matter:

```
┌───┐┌─┐      
│dot││x│      
└┬──┘└┬┘      
 │┌───▽──────┐
 ││reshape   │
 │└┬───────┬─┘
┌▽─▽──────┐│  
│pointwise││  
└┬────────┘│  
┌▽─────────▽┐ 
│convolution│ 
└───────────┘ 
```

After mlir fusion it just becomes: 

```
┌──────────────────────────┐┌─┐ 
│mlir_dot_reshape_pointwise││x│ 
└┬─────────────────────────┘└┬┘ 
 │┌──────────────────────────▽─┐
 ││reshape                     │
 │└┬───────────────────────────┘
┌▽─▽────────┐                   
│convolution│                   
└───────────┘ 
```

DCE wont remove the `reshape` since its used, and the duplicated `reshape` doesnt really matter since it is potentially an aliasing-like operator.